src/plan/aggregate.rs

```diff
@@ -43,94 +52,201 @@ pub struct Aggregate<P: Implementable> {
+                                }
                             }
                             // @TODO could preserve multiplicity of largest value here
                             output.push((*max, 1));
                         })
-                        .map(|(key, max)| key.iter().cloned()
-                             .chain([Value::Number(max)].iter().cloned())
-                             .collect())
-                }
-            },
-            AggregationFn::COUNT => {
-                SimpleRelation {
-                    symbols: self.variables.to_vec(),
-                    tuples: tuples,
-                        .map(|(ref _key, ref _tuple)| ())
-                        .count()
-                        .map(|(_, count)| vec![Value::Number(count as i64)])
+                        .map(|(key, max)| {
+                            let mut v = key.clone();
+                            v.push(Value::Number(max as i64));
+                            v
+                        }),
                 }
+            }
+            AggregationFn::COUNT => SimpleRelation {
+                symbols: self.variables.to_vec(),
+                tuples: tuples
+                    .consolidate()
+                    .group(|_key, input, output| {
+                        let count = input.len();
+                        output.push((count, 1))
+                    })
+                    .map(|(key, count)| {
+                        let mut v = key.clone();
+                        v.push(Value::Number(count as i64));
+                        v
+                    }),
             },
             AggregationFn::SUM => {
                 SimpleRelation {
                     symbols: self.variables.to_vec(),
                     tuples: tuples
+                        .consolidate()
+                        .distinct()
                         .explode(|(ref key, ref tuple)| {
                             let v = match tuple[0] {
                                 Value::Number(num) => num as isize,
-                                _ => panic!("COUNT can only be applied to numbers")
+                                _ => panic!("COUNT can only be applied to numbers"),
                             };
                             // @todo clone?
                             Some((key.clone(), v))
                         })
                         .consolidate()
                         .inner
-                        .map(|(data, time, delta)| (data.iter().cloned().chain([Value::Number(delta as i64)].iter().cloned()).collect(), time, delta))
-                        .as_collection()
+                        .map(|(data, time, delta)| {
+                            let mut v = data.clone();
+                            v.push(Value::Number(delta as i64));
+                            (v, time, delta)
+                        }).as_collection(),
                 }
             }
+            AggregationFn::AVG => SimpleRelation {
```

Comment:
I can blow your minds about average, if that would help! :D

Differential dataflow's `explode` operator (seeking a better name) allows you to move arbitrary data into the "count" part of a record. It needs to be a commutative group, but you can do for example:

```rust
tuples
    .map(prepare_unary)
    .explode(|(key, val)| Some((key, DiffPair::new(val, 1))))
    .count()
    .map(|(key, (sum, count))| (key, sum / count));
```

This has the often significant advantage that it maintains the average in a single `(key, (sum, cnt))` record, rather than using as many records as you have distinct `val`s for each key.

It's perhaps a bit weird and advanced, but I thought I would throw it out there in case that is interesting. `Variance` should have the same property, that it can be accumulated in place (explode to `(val * val, val, 1)`).

Reply:
Sweet. I'll have a look at it :). That seems quite elegant.

Reply:
Here is an example from TPC-H, query 1, where the goal is to sum and total a bunch of things:

https://github.com/frankmcsherry/differential-dataflow/blob/master/tpchlike/src/queries/query01.rs